Make unified broker accept list of plugins ID:version s and download meta.yaml from plugin registry

amisevsk opened the pull request:
### What does this PR do?
Adds capability for the unified broker to process a list of plugin id, version, and registry fields and download the `meta.yaml`s directly from registry instead of expecting the metas to be present in config.

This PR is broken up into three commits:
1. Add functionality for processing plugin FQNs (fully qualified names) and downloading meta.yamls
2. Modify `ioutil` to enable testing
3. Add tests for ioutil.

### Design decisions
- Support both plugin metas and plugin FQNs (i.e. it is backward compatible)
- Switch between processing metas and FQNs is with argument `--download-metas`
- Default registry can be specified via argument `--registry-address`
  - Default registry is optional but only allowed when all plugin FQNs specify a registry
- Expected config file in FQN case is a json list of objects with fields `id`, `version` and optionally `registry`. This was chosen to mirror `PluginFQN` as used in Che, as the list of `PluginFQN` can be easily serialized into the configmap for the broker.

### Testing directly
This PR also adds a file `brokers/unified/cmd/config-plugin-ids.json`. To test directly, the command
```
cd ./brokers/unified/cmd
go build main.go
./main \
  --disable-push \
  --runtime-id wsId:env:ownerId \
  --registry-address https://che-plugin-registry.openshift.io/plugins \
  --metas ./config-plugin_ids.json \
  --download-metas
```

### What issues does this PR fix or reference?
https://github.com/redhat-developer/rh-che/issues/1305

sleshchenko commented on utils/ioutil.go:
```diff
@@ -74,6 +96,55 @@ func (util *impl) Download(URL string, destPath string) error {
+
+		var pluginMeta model.PluginMeta
+		if err := yaml.Unmarshal(pluginRaw, &pluginMeta); err != nil {
+			return nil, fmt.Errorf("Failed to unmarshal downloaded meta.yaml: %s", err)
```
Please consider adding plugin id or even plugin URL to find broken meta.yaml easier.

amisevsk replied:
Added.

garagatyi commented on brokers/unified/cmd/config-plugin-ids.json:
```diff
@@ -0,0 +1,15 @@
+  {
+    "id": "org.eclipse.che.vscode-redhat.java-and-xml",
+    "version": "0.38.0",
+    "registry": "https://raw.githubusercontent.com/eclipse/che-plugin-registry/shareSidecarWithDebug/plugins/"
```
:+1:

amisevsk replied:
Will do -- I copy-pasted from the issue :)

amisevsk replied:
Changed to https://raw.githubusercontent.com/eclipse/che-plugin-registry/master/plugins/

garagatyi commented on brokers/unified/cmd/main.go:
```diff
@@ -34,6 +36,16 @@ func main() {
+		pluginFQNs := cfg.ParsePluginFQNs()
+		pluginMetas, err := utils.New().GetPluginMetas(pluginFQNs, cfg.RegistryAddress)
+		if err != nil {
+			log.Fatalf("Failed to download plugin metas: %s", err)
```
Broker has special method for this purpose that insures that error message is sent to WS master.

amisevsk replied:
I didn't realize -- done.

garagatyi commented on cfg/cfg.go:
```diff
@@ -101,6 +110,18 @@ func init() {
+		&DownloadMetas,
+		"download-metas",
+		false,
+		"Download plugin metadata from registry instead of process already-downloaded metas.",
```
Other messages don't have trailing slashes. Please, align new ones

amisevsk replied:
Not entirely sure what you mean, but I've removed `.` from the end of these messages.

garagatyi commented on cfg/cfg.go:
```diff
@@ -152,10 +173,35 @@ func Print() {
+
+	metas := make([]model.PluginMeta, 0)
+	if err := json.Unmarshal(raw, &metas); err != nil {
+		log.Fatal(err)
```
Broker has special method for this purpose that insures that error message is sent to WS master. I see that we exit program here including in all codebase. 
Consider changing this file to return error when something happens and then main.go of brokers would be able to send error to WS master

amisevsk replied:
This is a bit more involved of a change so it's broken out into a separate commit: https://github.com/eclipse/che-plugin-broker/pull/40/commits/b76456c6409d05e29010e1d0489228d717213949

In short: some of the broker implementations didn't include `common.Broker`, so to allow `broker.PrintFatal` to be called, I had to change interfaces and structs. Let me know what you think.

garagatyi replied:
It is nice that we now send error to WS master properly. Though I didn't get why we need to rename  `PushEvents` function. And it is better to remove main.go in other brokers instead of extending their codebases.

garagatyi replied:
I decided to remove them separately as this is left over from my last PR rather then something your PR specifically needs

amisevsk replied:
Thanks for that -- I'll rebase onto current master.

amisevsk replied:
Regarding the `PushEvents` function, that is something I'm currently working on a decent solution to; in e.g. the Theia broker, we're defining a `Broker` interface that includes a `PushEvents` function. To be able to use a Theia broker's `PrintFatal` we need to include `common.Broker` in this interface, which causes a name collision on the `PushEvents` function. However, since we're no longer using the non-unified brokers standalone, this can likely be removed.

garagatyi commented on model/model.go:
```diff
@@ -68,6 +68,12 @@ type PluginMeta struct {
 }
 
+type PluginFQN struct {
+	Registry string `json:"registry,omitempty" yaml:"registry,omitempty"`
```
Can you elaborate why we need omitempty here?

amisevsk replied:
"Need" is perhaps strong, but I included `omitempty` because I would expect `unmarshal -> marshal` to be idempotent, and we don't require the `registry` field in `PluginFQN` (i.e. from Che, it could be null and so omitted).

garagatyi replied:
I see. Thanks for the explanation

garagatyi commented on utils/ioutil.go:
```diff
@@ -38,6 +42,8 @@ type IoUtil interface {
 	Untar(tarPath string, dest string) error
 	CreateFile(file string, tr io.Reader) error
+	FetchJSON(url string) ([]byte, error)
+	GetPluginMetas(plugins []model.PluginFQN, defaultRegistry string) ([]model.PluginMeta, error)
```
Other methods in this file are business logic agnostic. Do you think we need to change the approach?

amisevsk replied:
I agree about business logic in `ioutil` to an extent, but the method is fairly coupled with `fetchJSON`, which is definitely a `ioutil` method. Moving it elsewhere would require creating a `ioutil` instance, which I think is somewhat clunky. 

I'm open to suggestions though.

amisevsk replied:
Also, at first glance, moving it out would make testing trickier, since mocking http.Client is necessary.

garagatyi replied:
We need to apply separation of concerns here. Util here is supposed to sugar cover boilerplate IO operations. Dealing with meta specific stuff is not supposed to be IO boilerplate code.
Thus I suggest moving code of downloading metas (or at least parsing them since we have download IO util function) to the broker. 
main.go of unified broker now also contains business logic and it wasn't supposed to.
Instead of that I would create `StartWithFQNs` function in addition to `Start` one and move business logic there. This should also help in avoiding changing other brokers (and you can avoid removing them as well).
Taking this into account main.go would do:
```go
if cfg.DownloadMetas {
    pluginFQNs, err := cfg.ParsePluginFQNs()
    if err ..... {}
    StartWithFQNs(pluginFQNs)
} else {
    pluginMetas, err := cfg.ReadConfig()
    if err ..... {}
    Start(pluginMetas)
}
```

Since we are going to remove `Start` function or make it do what `StartWithFQNs` does we should also deprecate it.

As for the unit testing VS Code broker tests contains cases with fetching something from remote server when remote server response mocking is needed. 
Here is function for that https://github.com/eclipse/che-plugin-broker/blob/b31e49ca87c87de4c251c7b0ffa6822682213612/brokers/test/utils.go#L139 that is used here https://github.com/eclipse/che-plugin-broker/blob/1451e5d1156bff4addb639050e848bd85969baca/brokers/vscode/broker_test.go#L65 in conjunction with https://github.com/eclipse/che-plugin-broker/blob/1451e5d1156bff4addb639050e848bd85969baca/brokers/vscode/broker_test.go#L877

amisevsk replied:
:+1: Makes sense, will do.

garagatyi commented on utils/ioutil.go:
```diff
@@ -46,10 +46,26 @@ type IoUtil interface {
+}
+
+// NewWithClient creates an instance of IoUtil using the provided client for all requests.
+func NewWithClient(client HttpClient) IoUtil {
```
Can you elaborate why do we need this function?

amisevsk replied:
This method is used to mock HTTPClient in the IoUtil tests; it makes it easier to test request failure, etc and makes the tests more robust. See utils/mocks/MockHttpClient.go

garagatyi replied:
There is a simpler mocking technique. See explanation how it is used in VS Code broker tests in one of my other comments

amisevsk commented on cfg/cfg.go:
```diff
@@ -101,6 +110,18 @@ func init() {
+		&RegistryAddress,
+		"registry-address",
+		"",
+		"Address of registry from which to retrieve meta.yamls when only plugin IDs are specified.",
```
Updated message, good point :+1:

amisevsk commented on utils/ioutil.go:
```diff
@@ -74,6 +96,55 @@ func (util *impl) Download(URL string, destPath string) error {
+			pluginURL = fmt.Sprintf("%s/%s/%s/meta.yaml", plugin.Registry, plugin.ID, plugin.Version)
+		} else {
+			if defaultRegistry == "" {
+				return nil, errors.New("plugin does not specify registry and no default is provided")
```
go-lint suggests: `error strings should not be capitalized or end with punctuation or a newline`

Updated message to include plugin ID.

amisevsk replied:
(The reasoning behind this is that often error messages are combined: `fmt.Sprintf("could not do x: %s", err)`, so adding capitalization introduces caps halfway through the message.)

amisevsk commented on utils/ioutil.go:
```diff
@@ -74,6 +96,55 @@ func (util *impl) Download(URL string, destPath string) error {
+		}
+		pluginRaw, err := util.FetchJSON(pluginURL)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to fetch plugin meta.yaml from registry: %s", err)
```
:+1: Done

garagatyi commented:
If you don't want to delete them, you need to avoid changes in them.

garagatyi commented:
@amisevsk I'm working on another issue that is related to this PR. Do you have draft changes on WS master side to use broker with these changes? I am asking because you could have them to test your changes integrated in Che and it would be better if I do not overlap with your changes or would be able to identify if those changes I wanted to do are compatible with yours

amisevsk commented:
> Do you have draft changes on WS master side to use broker with these changes?

I don't yet, but should have something today.

amisevsk commented:
@garagatyi I've rebased onto master and added two commits:

- https://github.com/eclipse/che-plugin-broker/pull/40/commits/7ebbb2874fecfd6fdadcb2282b00765951e17609 is a minor refactor to move the broker interface to the common package (since there are three identical interfaces in each of che plugin, theia, and vscode brokers)

- https://github.com/eclipse/che-plugin-broker/pull/40/commits/f9f828a72f09af51dd19180a5d41fc801f4871a5 does the requested changes: it
  - Moves meta downloading to unified broker.
  - Reworks tests as requested.

PTAL

garagatyi commented on brokers/unified/broker.go:
```diff
@@ -125,10 +141,41 @@ func (b *Broker) ProcessPlugins(metas []model.PluginMeta) error {
+			}
+			registry = strings.TrimSuffix(defaultRegistry, "/")
+		}
+		pluginURL := fmt.Sprintf("%s/%s/%s/meta.yaml", registry, plugin.ID, plugin.Version)
```
Taking into account that Che properties specify plugin registry without `/plugins` suffix (`che.workspace.plugin_registry_url=https://che-plugin-registry.openshift.io`) it would be better to align them and add `/plugins` here too. Treat it as API, not just file hosting.

garagatyi replied:
Since in a case of old plugin `/plugins` can be treated as `publisher` we probbaly can not avoid setting plugin registry without `/plugins`

amisevsk replied:
I've pulled the URL format into a const at the top, but now I'm thinking whether it makes sense to include it in the config at some point. There would also be the problem of using multiple registries for different plugins; they would all have to conform to our current spec which could be tricky.

garagatyi commented on brokers/che-plugin-broker/broker.go:
```diff
@@ -43,30 +43,23 @@ const (
 )
 
-// ChePluginBroker is used to process Che plugins
-type ChePluginBroker interface {
```
@davidfestal this is something that can affect your work

amisevsk replied:
Since they're all the same, you should be able to use `common.BrokerImpl` instead.
